### PR TITLE
[PIP] Implement cards x4

### PR DESCRIPTION
**Pull request by ssk97**
Implementing more uncommons from Fallout

### CODE REVIEW

File: Mage.Sets/src/mage/cards/i/IanTheReckless.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    public IanTheReckless(UUID ownerId, CardSetInfo setInfo) {
+        super(ownerId, setInfo, new CardType[]{CardType.CREATURE}, "{1}{R}");
+
+        this.supertype.add(SuperType.LEGENDARY);
+        this.subtype.add(SubType.HUMAN);
+        this.subtype.add(SubType.WARRIOR);
+        this.power = new MageInt(2);
+        this.toughness = new MageInt(1);
+
+        // Whenever Ian the Reckless attacks, if it's modified, you may have it deal damage equal to its power to you and any target.
```
Comment (JayDi85) [inline Mage.Sets/src/mage/cards/i/IanTheReckless.java:45]: Looks like any cards with `ModifiedPredicate` usage can have `ConditionHint` for better UX.

File: Mage.Sets/src/mage/cards/l/LegateLaniusCaesarsAce.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    }
+
+    @Override
+    public LegateLaniusCaesarsAce copy() {
+        return new LegateLaniusCaesarsAce(this);
+    }
+}
+
+//Based on SacrificeAllEffect
+class LegateLaniusCaesarsAceSacrificeEffect extends OneShotEffect {
```
Comment (JayDi85) [inline Mage.Sets/src/mage/cards/l/LegateLaniusCaesarsAce.java:59]: Try to use standard effect `SacrificeOpponentsEffect` with `HalfValue` dynamic values instead custom. Also you can add card hint with that value like "You would have to sacrifice: xxx"

File: Mage.Sets/src/mage/cards/l/LegateLaniusCaesarsAce.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+            }
+            // 1/10 rounded up
+            int num = (game.getBattlefield().count(StaticFilters.FILTER_PERMANENT_A_CREATURE,playerId,source, game)+9)/10;
+            int numTargets = Math.min(num, game.getBattlefield().count(TargetSacrifice.makeFilter(StaticFilters.FILTER_PERMANENT_A_CREATURE), player.getId(), source, game));
+            if (numTargets < 1) {
+                continue;
+            }
+            TargetSacrifice target = new TargetSacrifice(numTargets, StaticFilters.FILTER_PERMANENT_A_CREATURE);
+            while (!target.isChosen() && target.canChoose(playerId, source, game) && player.canRespond()) {
+                player.choose(Outcome.Sacrifice, target, source, game);
```
Comment (JayDi85) [inline Mage.Sets/src/mage/cards/l/LegateLaniusCaesarsAce.java:90]: AI can freeze here due wrong or cancelled choice.

Code protection for AI freeze:
- stop cycle on false choose (required for any "while" code);
- remove chosen value from a list for next choice by filter or cards list modification (recommended if possible).

So must add break here.

**Review by JayDi85 — CHANGES_REQUESTED** [review]

File: Mage.Sets/src/mage/cards/l/LegateLaniusCaesarsAce.java (hunk 3 above)
Comment (ssk97) [reply Mage.Sets/src/mage/cards/l/LegateLaniusCaesarsAce.java:90]: This is currently exactly copied from the existing `SacrificeAllEffect`, and is also identical to `SacrificeEffect`'s. I'm not sure why this is a loop at all, honestly.

File: Mage.Sets/src/mage/cards/l/LegateLaniusCaesarsAce.java (hunk 2 above)
Comment (ssk97) [reply Mage.Sets/src/mage/cards/l/LegateLaniusCaesarsAce.java:59]: The current `SacrificeOpponentsEffect` calculates the number once at the start. I could move it to inside the loop, but the dynamic value's calculate function still has no good way of knowing which opponent is being evaluated. I suppose I could smuggle the information in as a local variable, but that felt like a sufficiently messy hack that it shouldn't be added to the common class for just one card.

File: Mage.Sets/src/mage/cards/i/IanTheReckless.java (hunk 1 above)
Comment (ssk97) [reply Mage.Sets/src/mage/cards/i/IanTheReckless.java:45]: That's generally used for "if you control a modified creature".

File: Mage.Sets/src/mage/cards/l/LegateLaniusCaesarsAce.java (hunk 3 above)
Comment (xenohedron) [reply Mage.Sets/src/mage/cards/l/LegateLaniusCaesarsAce.java:90]: I tried to remove the while loop when I refactored as part of #11587 but it caused a test failure, BarterInBloodTest I believe. I didn't investigate further why the choose method needs a loop to select multiple, seems not ideal. But it does serve a purpose.

File: Mage.Sets/src/mage/cards/l/LegateLaniusCaesarsAce.java (hunk 2 above)
Comment (xenohedron) [reply Mage.Sets/src/mage/cards/l/LegateLaniusCaesarsAce.java:59]: Agree that custom class is fine here

File: Mage.Sets/src/mage/cards/i/IanTheReckless.java (hunk 1 above)
Comment (xenohedron) [reply Mage.Sets/src/mage/cards/i/IanTheReckless.java:45]: yeah, I don't think a hint is really needed here

File: Mage.Sets/src/mage/cards/l/LegateLaniusCaesarsAce.java (hunk 3 above)
Comment (JayDi85) [reply Mage.Sets/src/mage/cards/l/LegateLaniusCaesarsAce.java:90]: `BarterInBloodTest` uses AI targeting instead direct test commands and it fails on "while" remove. Looks like AI do not support multiple targets in choice dialogs (on resolve) and it works fine for cast/activate only (cause it uses game simulations for it).

Good code from human:
shot_240402_070836

Bad code from computer:
shot_240402_071218

It's a more global problem with AI, not related to PR.

File: Mage.Sets/src/mage/cards/i/IanTheReckless.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        TriggeredAbility ability = new AttacksTriggeredAbility(new DamageControllerEffect(
+                new SourcePermanentPowerCount()).setText("have it deal damage equal to its power to you"), true);
+        ability.addEffect(new DamageTargetEffect(new SourcePermanentPowerCount()).setText("and any target"));
+        ability.addTarget(new TargetAnyTarget());
+        this.addAbility(new ConditionalInterveningIfTriggeredAbility(ability, condition,
+                "Whenever Ian the Reckless attacks, if it's modified, you may have it deal damage equal to its power to you and any target."));
```
Comment (xenohedron) [inline Mage.Sets/src/mage/cards/i/IanTheReckless.java:51]: {this}

File: Mage.Sets/src/mage/cards/g/GunnerConscript.java [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+        this.power = new MageInt(2);
+        this.toughness = new MageInt(2);
+
+        // Trample
+        this.addAbility(TrampleAbility.getInstance());
+
+        // Gunner Conscript gets +1/+1 for each Aura and Equipment attached to it.
+        DynamicValue auraAmount = new AuraAttachedCount();
+        DynamicValue equipAmount = new EquipmentAttachedCount();
+        DynamicValue totalAmount = new AdditiveDynamicValue(auraAmount, equipAmount);
```
Comment (xenohedron) [inline Mage.Sets/src/mage/cards/g/GunnerConscript.java:48]:
```java
private static final DynamicValue totalAmount = new AdditiveDynamicValue(new AuraAttachedCount(), new EquipmentAttachedCount());
```

**Review by xenohedron — APPROVED** [review]
I think it's fine to merge this, and track the while loop thing separately with #12044

**Pull request merged**